cram_size.c

```diff
@@ -0,0 +1,656 @@
+
+            // Embedded reference.  Check it's consistent (if used this is
+            // an almost guaranteed certainty, so we take the easy route).
+            if (ref_seq_blk >= 0) {
+                if (ref_seq_blk_used == -1)
+                    ref_seq_blk_used = ref_seq_blk;
+                else if (ref_seq_blk_used != ref_seq_blk)
+                    fprintf(stderr, "Embedded reference is not consistently using the same Content-Id.\n"
+                            "Reported figures for reference will be invalid.\n");
+            }
+
+            // Slice data blocks
+            for (j = 0; j < num_blocks; j++) {
+                // read and discard, unless it's the ref-ID block
+                if (!(blk = cram_read_block(in_c)))
+                    goto err;
+
+                int32_t csize = cram_block_get_comp_size(blk);
+                int32_t usize = cram_block_get_uncomp_size(blk);
+                int cid = cram_block_get_content_id(blk);
+                enum cram_block_method method = cram_block_get_method(blk);
+
+                // Expand comp to the internal sub-formats, eg
+                // rANS order-0/1, PACK+RLE, etc.
+                cram_method_details *cm;
+                cm = cram_expand_method(cram_block_get_data(blk),
+                                        cram_block_get_comp_size(blk),
+                                        method);
+                if (!cm)
+                    goto err;
+                enum comp_expanded comp
+                    = comp_method2expanded(cm);
+                free(cm);
+
+                k = kh_put(cu, cu_size, cid, &ret);
+                if (ret < 0)
+                    goto err;
+                if (ret == 0) {
+                    kh_value(cu_size, k).csize[comp] += csize;
+                    kh_value(cu_size, k).usize[comp] += usize;
+                } else {
+                    memset(&kh_value(cu_size, k), 0, sizeof(cusize_t));
+                    kh_value(cu_size, k).csize[comp]  = csize;
+                    kh_value(cu_size, k).usize[comp]  = usize;
+                }
+
+                cram_free_block(blk);
+                blk = NULL;
+            }
+            cram_free_slice_header(shdr);
+            shdr = NULL;
+        }
+
+        cram_free_container(c);
+        c = NULL;
+    }
+
+    off_t tot_size = report_size(outfp, verbose, ref_seq_blk_used,
+                                 cu_size, cid2ds);
+    if (tot_size < 0)
+        goto err;
+
+    kh_destroy(cu, cu_size);
+    cram_cid2ds_free(cid2ds);
+
+    // This is a little bit naughty as it assumes that the hFILE pointer
+    // is the very first thing in the opaque cram_fd struct, which it is.
+    // The alternative would be to add an explicit htslib CRAM API member
+    // function to return the internal hFILE handle, but if we expose this
+    // then people may feel it's fair game to start modifying it.
+    off_t end = htell(*(hFILE **)in->fp.cram);
```

Comment:
This naughtiness could be avoided by opening the input using `hopen()` and then using `hts_hopen()` to upgrade it to a `samFile`.  That way you would still have a pointer to the underlying `hFILE` to pass to `htell()`.

Reply:
Oh that's useful!  I was wondering how to get an hfile out and was rather surprised it wasn't possible, but this trick avoids the necessity.